programs/bpf_loader/src/serialization.rs

```diff
@@ -99,59 +120,55 @@ impl Serializer {
+                // put the realloc padding in its own region
+                self.push_region(account.can_data_be_changed().is_ok());
             }
         }
 
-        Ok(())
+        Ok(vm_data_addr)
     }
 
-    fn push_account_region(
+    fn push_account_data_region(
         &mut self,
         account: &mut BorrowedAccount<'_>,
     ) -> Result<(), InstructionError> {
-        self.push_region();
-        let account_len = account.get_data().len();
-        if account_len > 0 {
-            let region = if account.can_data_be_changed().is_ok() {
-                if account.is_shared() {
-                    // If the account is still shared it means it wasn't written to yet during this
-                    // transaction. We map it as CoW and it'll be copied the first time something
-                    // tries to write into it.
-                    let index_in_transaction = account.get_index_in_transaction();
-
-                    MemoryRegion::new_cow(
-                        account.get_data(),
-                        self.vaddr,
-                        index_in_transaction as u64,
-                    )
-                } else {
-                    // The account isn't shared anymore, meaning it was written to earlier during
-                    // this transaction. We can just map as writable no need for any fancy CoW
-                    // business.
+        if !account.get_data().is_empty() {
+            let region = match account_data_region_memory_state(account) {
+                MemoryState::Readable => MemoryRegion::new_readonly(account.get_data(), self.vaddr),
+                MemoryState::Writable => {
                     MemoryRegion::new_writable(account.get_data_mut()?, self.vaddr)
                 }
-            } else {
-                MemoryRegion::new_readonly(account.get_data(), self.vaddr)
+                MemoryState::Cow(index_in_transaction) => {
+                    MemoryRegion::new_cow(account.get_data(), self.vaddr, index_in_transaction)
+                }
             };
             self.vaddr += region.len;
             self.regions.push(region);
         }
+
         Ok(())
     }
 
-    fn push_region(&mut self) {
+    fn push_region(&mut self, writable: bool) {
         let range = self.region_start..self.buffer.len();
-        let region = MemoryRegion::new_writable(
-            self.buffer.as_slice_mut().get_mut(range.clone()).unwrap(),
-            self.vaddr,
-        );
+        let region = if writable {
+            MemoryRegion::new_writable(
+                self.buffer.as_slice_mut().get_mut(range.clone()).unwrap(),
+                self.vaddr,
+            )
+        } else {
+            MemoryRegion::new_readonly(
+                self.buffer.as_slice().get(range.clone()).unwrap(),
+                self.vaddr,
+            )
+        };
         self.regions.push(region);
         self.region_start = range.end;
         self.vaddr += range.len() as u64;
```

Comment:
nit: i'd prefer this reordering to be consistent with `push_account_region()`:

```rust
        self.vaddr += range.len() as u64;
        self.regions.push(region);
        self.region_start = range.end; // put this line at end because this is a state mutation for _next_ region.


```